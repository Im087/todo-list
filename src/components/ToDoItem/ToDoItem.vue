<template>
  <v-container class="to-do-item d-flex flex-wrap bg-grey-lighten-3 rounded mb-2">
    <v-checkbox
      class="flex-grow-0 to-do-item__checkbox"
      :class="toDoItem.itemStatus ? 'to-do-item__checkbox--checked' : 'to-do-item__checkbox--unchecked'"
      :label="toDoItem.itemStatus ? 'Done' : 'Undone'"
      color="success"
      hide-details
      v-model="toDoItem.itemStatus"
      data-test-id="to-do-item__checkbox"
    >
    </v-checkbox>
    <v-textarea
      class="flex-grow-1"
      rows="1"
      bg-color="transparent"
      :disabled="toDoItem.isDisabled"
      v-model="toDoItem.toDoContent"
      hide-details
      clearable
      no-resize
      data-test-id="to-do-item__textarea"
    >
    </v-textarea>
    <div class="ml-4">
      <v-btn
        v-if="toDoItem.isDisabled"
        class="mt-2 mr-2"
        color="warning"
        size="large"
        @click="editToDoItem()"
        data-test-id="to-do-item__edit-btn"
      >
        Edit
      </v-btn>
      <v-btn
        v-if="!toDoItem.isDisabled"
        class="mt-2 mr-2"
        color="success"
        size="large"
        @click="confirmToDoItem()"
        data-test-id="to-do-item__confirm-btn"
      >
        OK
      </v-btn>
      <v-btn
        class="mt-2"
        color="error"
        size="large"
        @click="deleteToDoItem(toDoItem.timeStamp)"
        data-test-id="to-do-item__delete-btn"
      >
        Delete
      </v-btn>
    </div>
  </v-container>
</template>

<script lang="ts">
import { defineComponent, toRefs, type PropType } from "vue";
import type { ToDoItem } from "@/interfaces";
import store from "@/store";

export default defineComponent ({
  name: "ToDoItem",
  props: {
    toDoItem: {
      type: Object as PropType<ToDoItem>,
      required: true,
      deep: true
    }
  },
  setup(props) {
    const toDoItem = toRefs(props).toDoItem;

    const editToDoItem = () => {
      toDoItem.value.isDisabled = false;
    };
    const confirmToDoItem = () => {
      toDoItem.value.isDisabled = true;
    };
    const deleteToDoItem = (timeStamp: string) => {
      store.dispatch('deleteToDoItem', timeStamp);
    };

    return {
      toDoItem,
      editToDoItem,
      confirmToDoItem,
      deleteToDoItem
    };
  }
});
</script>

<style lang="scss" scoped>
.to-do-item {

  &__checkbox {
    min-width: 100px;

    &--checked :deep(.v-label) {
      color: green;
    }

    &--unchecked :deep(.v-label) {
      color: red;
    }
  }
}
</style>