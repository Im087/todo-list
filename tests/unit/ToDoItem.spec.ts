import { describe, test, expect, vi, beforeEach, afterEach } from "vitest";
import ToDoItem from '../../src/components/ToDoItem/ToDoItem.vue';
import { VueWrapper, mount } from '@vue/test-utils';
import { createVuetify } from "vuetify";
import * as components from "vuetify/components";
import * as directives from "vuetify/directives";

describe('ToDoItem.vue', () => {
  const vuetify = createVuetify({components, directives});
  const $store = {
    state: {
      toDoArray: []
    },
    dispatch: vi.fn(),
    commit: vi.fn()
  };

  describe('When mounted', () => {
    const toDoItem = {
      itemStatus: false,
      isDisabled: true,
      timeStamp: new Date().toISOString(),
      toDoContent: 'todo',
    };
    const wrapper = mount(ToDoItem, {
      global: {
        plugins: [vuetify]
      },
      props: {
        toDoItem
      }
    });

    test('get props', () => {
      expect(wrapper.props().toDoItem).toEqual(toDoItem);
    })

    wrapper.unmount();
  });

  describe('When checkbox checked', async () => {
    const toDoItem = {
      itemStatus: false,
      isDisabled: true,
      timeStamp: new Date().toISOString(),
      toDoContent: 'todo',
    };
    const wrapper = mount(ToDoItem, {
      global: {
        plugins: [vuetify]
      },
      props: {
        toDoItem
      },
      attachTo: document.body
    });
    const vCheckbox = wrapper.getComponent('[data-test-id="to-do-item__checkbox"]');

    await vCheckbox.get('input').trigger('click');

    test('the checked attribute is true', () => {
      expect(vCheckbox.get('input').element.checked).toBeTruthy();
    });
    test('toDoItem.itemStatus is true', () => {
      expect(wrapper.vm.toDoItem.itemStatus).toBeTruthy();
    });
    test('element class is .to-do-item__checkbox--checked', () => {
      expect(vCheckbox.classes()).toContain('to-do-item__checkbox--checked');
    });
    test('label text will be Done', () => {
      expect(vCheckbox.text()).toEqual('Done');
    });

    wrapper.unmount();
  });

  describe('When checkbox unchecked', async () => {
    const toDoItem = {
      itemStatus: true,
      isDisabled: true,
      timeStamp: new Date().toISOString(),
      toDoContent: 'todo',
    };
    const wrapper = mount(ToDoItem, {
      global: {
        plugins: [vuetify]
      },
      props: {
        toDoItem
      },
      attachTo: document.body
    });
    const vCheckbox = wrapper.getComponent('[data-test-id="to-do-item__checkbox"]');

    await vCheckbox.get('input').trigger('click');

    test('the checked attribute is false', () => {
      expect(vCheckbox.get('input').element.checked).toBeFalsy();
    });
    test('toDoItem.itemStatus is false', () => {
      expect(wrapper.vm.toDoItem.itemStatus).toBeFalsy();
    });
    test('element class is .to-do-item__checkbox--checked', () => {
      expect(vCheckbox.classes()).toContain('to-do-item__checkbox--unchecked');
    });
    test('label text will be Done', () => {
      expect(vCheckbox.text()).toEqual('Undone');
    });

    wrapper.unmount();
  });

  describe('When clicking on EDIT button', async () => {
    const toDoItem = {
      itemStatus: false,
      isDisabled: true,
      timeStamp: new Date().toISOString(),
      toDoContent: 'todo',
    };
    const wrapper = mount(ToDoItem, {
      global: {
        plugins: [vuetify]
      },
      props: {
        toDoItem
      }
    });
    const vBtn = wrapper.getComponent('[data-test-id="to-do-item__edit-btn"]');
    const spy = vi.spyOn(wrapper.vm, 'editToDoItem');
    
    await vBtn.get('button').trigger('click');

    test('editToDoItem function is called', () => {
      expect(spy).toHaveBeenCalled();
    });
    test('textarea is activated', () => {
      expect(wrapper.findComponent('[data-test-id="to-do-item__textarea"]').get('textarea').element.disabled).toBeFalsy();
    });
    test('toToItem.isDisabled is false', () => {
      expect(wrapper.vm.toDoItem.isDisabled).toBeFalsy();
    });
    // test('EDIT button is destroyed', () => {
    //   expect(wrapper.findComponent('[data-test-id="to-do-item__edit-btn"]').exists()).toBeFalsy();
    // });
    // test('OK button is created', () => {
    //   expect(wrapper.findComponent('[data-test-id="to-do-item__confirm-btn"]').exists()).toBeTruthy();
    // });

    wrapper.unmount();
  });

  describe('When clicking on OK button', async () => {
    const toDoItem = {
      itemStatus: false,
      isDisabled: false,
      timeStamp: new Date().toISOString(),
      toDoContent: 'todo',
    };
    const wrapper = mount(ToDoItem, {
      global: {
        plugins: [vuetify]
      },
      props: {
        toDoItem
      }
    });
    const vBtn = wrapper.getComponent('[data-test-id="to-do-item__confirm-btn"]');
    const spy = vi.spyOn(wrapper.vm, 'confirmToDoItem');

    await vBtn.get('button').trigger('click');

    test('confirmToDoItem function is called', () => {
      expect(spy).toBeCalled();
    });
    test('textarea is disabled', () => {
      expect(wrapper.findComponent('[data-test-id="to-do-item__textarea"]').get('textarea').element.disabled).toBeTruthy();
    });
    test('toToItem.isDisabled is true', () => {
      expect(wrapper.vm.toDoItem.isDisabled).toBeTruthy();
    });
    // test('OK button is destroyed', () => {
    //   expect(wrapper.findComponent('[data-test-id="to-do-item__confirm-btn"]').exists()).toBeFalsy();
    // });
    // test('EDIT button is created', () => {
    //   expect(wrapper.findComponent('[data-test-id="to-do-item__edit-btn"]').exists()).toBeTruthy();
    // });

    wrapper.unmount();
  });

  describe('When clicking on DELETE button', async () => {
    const toDoItem = {
      itemStatus: false,
      isDisabled: true,
      timeStamp: new Date().toISOString(),
      toDoContent: 'todo',
    };
    const wrapper = mount(ToDoItem, {
      global: {
        plugins: [vuetify]
      },
      props: {
        toDoItem
      },
      mocks: {
        $store
      }
    });
    const vBtn = wrapper.getComponent('[data-test-id="to-do-item__delete-btn"]');
    const spy = vi.spyOn(wrapper.vm, 'deleteToDoItem');

    await vBtn.get('button').trigger('click');

    test('deleteToDoItem is called', () => {
      expect(spy).toBeCalled();
    });
    // test('vuex action is called', () => {
    //   expect($store.dispatch).toBeCalled();
    // })

    wrapper.unmount();
  });

  describe('When the textarea updates', async () => {
    const toDoItem = {
      itemStatus: false,
      isDisabled: false,
      timeStamp: new Date().toISOString(),
      toDoContent: 'todo',
    };
    const wrapper = mount(ToDoItem, {
      global: {
        plugins: [vuetify]
      },
      props: {
        toDoItem
      }
    });
    const vTextarea = wrapper.getComponent('[data-test-id="to-do-item__textarea"]');

    await vTextarea.get('textarea').setValue('new todo');

    test('two-way binding will work', () => {
      expect(wrapper.vm.toDoItem.toDoContent).toEqual('new todo');
    });

    wrapper.unmount();
  });

  describe('Snapshot', () => {
    const toDoItem = {
      itemStatus: false,
      isDisabled: true,
      timeStamp: new Date().toISOString(),
      toDoContent: 'todo',
    };
    const wrapper = mount(ToDoItem, {
      global: {
        plugins: [vuetify]
      },
      props: {
        toDoItem
      }
    });

    test('is OK', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    wrapper.unmount();
  });
});