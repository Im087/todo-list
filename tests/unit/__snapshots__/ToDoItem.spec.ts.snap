// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToDoItem.vue > Snapshot > is OK 1`] = `
<div
  class="v-container to-do-item d-flex flex-wrap bg-grey-lighten-3 rounded mb-2"
  data-v-4d72f269=""
>
  <div
    class="v-input v-input--horizontal v-input--density-default v-input--dirty v-checkbox flex-grow-0 to-do-item__checkbox to-do-item__checkbox--unchecked"
    data-test-id="to-do-item__checkbox"
    data-v-4d72f269=""
  >
    <!---->
    <div
      class="v-input__control"
    >
      
      <div
        class="v-selection-control v-selection-control--inline v-selection-control--density-default v-checkbox-btn"
      >
        <div
          class="v-selection-control__wrapper"
        >
          <!---->
          <div
            class="v-selection-control__input"
          >
            <i
              aria-hidden="true"
              class="mdi-checkbox-blank-outline mdi v-icon notranslate v-theme--light v-icon--size-default"
            />
            <input
              aria-describedby="checkbox-0-messages"
              aria-disabled="false"
              id="checkbox-0"
              indeterminateicon="$checkboxIndeterminate"
              type="checkbox"
              value="true"
            />
            <!---->
          </div>
        </div>
        <label
          class="v-label v-label--clickable"
          for="checkbox-0"
        >
          <!---->
          
          Undone
          
        </label>
      </div>
      
    </div>
    <!---->
    <!---->
  </div>
  <div
    class="v-input v-input--horizontal v-input--density-default v-input--dirty v-input--disabled v-textarea v-text-field v-textarea--no-resize flex-grow-1"
    data-test-id="to-do-item__textarea"
    data-v-4d72f269=""
  >
    <!---->
    <div
      class="v-input__control"
    >
      
      <div
        class="v-field v-field--active v-field--appended v-field--disabled v-field--dirty v-field--has-background v-field--no-label v-field--variant-filled v-theme--light bg-transparent"
        role="textbox"
      >
        <div
          class="v-field__overlay"
        />
        <div
          class="v-field__loader"
        >
          <div
            aria-hidden="true"
            aria-valuemax="100"
            aria-valuemin="0"
            class="v-progress-linear v-theme--light"
            role="progressbar"
            style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);"
          >
            <!---->
            <div
              class="v-progress-linear__background"
              style="width: 100%;"
            />
            <transition-stub
              appear="false"
              css="true"
              name="fade-transition"
              persisted="false"
            >
              <div
                class="v-progress-linear__indeterminate"
              >
                
                <div
                  class="v-progress-linear__indeterminate long"
                />
                <div
                  class="v-progress-linear__indeterminate short"
                />
                
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div
          class="v-field__field"
          data-no-activator=""
        >
          <!---->
          <label
            class="v-label v-field-label"
            for="input-4"
          >
            <!---->
            
            <!---->
            
          </label>
          
          
          <!---->
          <textarea
            aria-describedby="input-4-messages"
            class="v-field__input"
            disabled=""
            id="input-4"
            rows="1"
          />
          <!---->
          <!---->
          
          
        </div>
        <transition-stub
          appear="false"
          css="true"
          name="expand-x-transition"
          persisted="false"
        >
          <div
            class="v-field__clearable"
          >
            <i
              aria-hidden="false"
              aria-label="Clear "
              class="mdi-close-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable"
              role="button"
            />
          </div>
        </transition-stub>
        <div
          class="v-field__append-inner"
        >
          <!---->
          <!---->
        </div>
        <div
          class="v-field__outline"
        >
          <!---->
          <!---->
        </div>
      </div>
      
    </div>
    <!---->
    <!---->
  </div>
  <div
    class="ml-4"
    data-v-4d72f269=""
  >
    <button
      class="v-btn v-btn--elevated v-theme--light bg-warning v-btn--density-default v-btn--size-large v-btn--variant-elevated mt-2 mr-2"
      data-test-id="to-do-item__edit-btn"
      data-v-4d72f269=""
      type="button"
    >
      
      <span
        class="v-btn__overlay"
      />
      <span
        class="v-btn__underlay"
      />
      
      <!---->
      <span
        class="v-btn__content"
        data-no-activator=""
      >
        
         Edit 
        
      </span>
      <!---->
      <!---->
    </button>
    <!--v-if-->
    <button
      class="v-btn v-btn--elevated v-theme--light bg-error v-btn--density-default v-btn--size-large v-btn--variant-elevated mt-2"
      data-test-id="to-do-item__delete-btn"
      data-v-4d72f269=""
      type="button"
    >
      
      <span
        class="v-btn__overlay"
      />
      <span
        class="v-btn__underlay"
      />
      
      <!---->
      <span
        class="v-btn__content"
        data-no-activator=""
      >
        
         Delete 
        
      </span>
      <!---->
      <!---->
    </button>
  </div>
</div>
`;
